proposals/0258-property-wrappers.md

```diff
@@ -485,6 +485,18 @@ enum MemoryOrder {
 };
 ```
 
+The Atomic property wrapper is class rather than a struct because the
+memory guarded by synchronization primitives must be independent from
+the wrapper. The wrapper's value will be loaded before the call to
+`wrappedValue.getter` and written back after the call to
+`wrappedValue.setter`. Therefore, synchronization within the wrapper
+cannot provide atomic access to its own value. However, when the
+wrapped value is a separate stored property within the wrapper object,
+then it is accessed independently, only after calling the
+wrappedValue's getter and setter. Note that a class type property
+wrapper gives the wrapped value reference semantics. All copies of the
+parent object will share the same atomic value.
```

Comment:
Suggestion:

```
The `Atomic` property wrapper is a class rather than a struct because the memory guarded by synchronization primitives must be independent from the wrapper. In Swift's formal memory access model, methods on a value types are considered to access the entire value, and so calling the `wrappedValue` getter formally reads the entire stored wrapper, while calling the setter of `wrappedValue` formally modifies the entire stored wrapper. This happens outside of the wrapper implementation, and the wrapper has no ability to affect it. If this formal access actually causes the memory to be copied, the atomicity of the operation will be broken. Because of this, using a `struct` is not semantically correct, and tools such as Thread Sanitizer will correctly report races when the atomic is used concurrently. Methods on classes, in contrast, are just passed a reference to `self` and do not automatically formally access its memory, so they can internally ensure the memory is only accessed using appropriate atomics. Using a class wrapper does mean that the wrapped value will have reference semantics and so different copies of it will share the same atomic storage. Allowing an atomic wrapper with value semantics is an interesting future direction.
```